### PR TITLE
AS-603

### PR DESCRIPTION
**Pull request by sagehen03**
Greetings! This is my first pass at https://broadworkbench.atlassian.net/browse/AS-602

- ~~There are currently 3 GCF implementations (DeltaLayerCloudFunction, DeltaLayerBackgroundFunction, DeltaLayerRawFunction) we need to decide on which we prefer.  All of them delegate to the same code to carry out the business logic. My vote is for DeltaLayerRawFunction.~~ DeltaLayerRawFunction has been selected.
- There are two DeltaLayerBigQueryWriter implementations.  We should eventually chose one though we could defer that choice until after we conduct some performance tests.
- There's currently no retry logic or robust error handling for fetching the file from cloud storage or for writing it's contents to BQ.  I think both Google libraries we're using for those purposes might already support retries.
- I don't think my names are particularly good.  I'd like to get some suggestions from folks who understand delta layer's terminology a bit better.

### CODE REVIEW

File: .github/actions/build-and-deploy-backend-function/action.yml
Comment (sagehen03) [inline .github/actions/build-and-deploy-backend-function/action.yml:50]: This is a temporary change.  We will re-enable github actions to deploy functions soon.

File: src/main/java/bio/terra/cloudfiletodatastore/FileCreatedMessageHarness.java
Comment (sagehen03) [inline src/main/java/bio/terra/cloudfiletodatastore/FileCreatedMessageHarness.java:1]: As it stands currently, I don't think the code needs this interface.

File: src/main/java/bio/terra/cloudfiletodatastore/FileUploadedMessage.java
Comment (sagehen03) [inline src/main/java/bio/terra/cloudfiletodatastore/FileUploadedMessage.java:1]: I'm sure we'll need to add more information to this class especially if it's shared between TestRunner and DL

File: src/main/java/bio/terra/cloudfiletodatastore/ResourceFetcher.java
Comment (sagehen03) [inline src/main/java/bio/terra/cloudfiletodatastore/ResourceFetcher.java:1]: We need to figure out whether using the slightly more unwieldy InputStream approach is worth the memory savings--something we'll probably need to load test.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBigQueryWriter.java
Comment (sagehen03) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBigQueryWriter.java:1]: I'm using the SQL-like BQ API in this class and in DeltaLayerBQInsertGenerator since it features prominently in Google's doc and is fairly intuitive, but there are two alternatives which may offer advantages.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBqInsertGenerator.java
Comment (sagehen03) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBqInsertGenerator.java:55]: I need to look into how we format dates for point correction files.

**Review by davidangb — COMMENTED** [review]
here are a few comments so far, I am 1/3 of the way through review and will keep going!

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/DeltaLayerBackgroundFunction.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/DeltaLayerBackgroundFunction.java:16]: does the crummy gson support throw an error if our copy of `GCSEvent` leaves out all the fields we don't care about? For instance, if we removed `kmsKeyName` from `GCSEvent`, would it cause a problem? I like the `DeltaLayerBackgroundFunction` approach better than the other two, IF we can keep our maintenance of `GCSEvent` low.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/model/json/PointCorrectionDestination.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/model/json/PointCorrectionDestination.java:62]: here again, looking at lombok or immutable or something that helps us with boilerplate

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/model/json/PointCorrectionOperation.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/model/json/PointCorrectionOperation.java:11]: in the json files that exist in GCS, these won't always be strings. The values here will be any json datatype, so you might see:
```
"value": "my_new_value"
```
for strings, 
```
"value": 123.45
```
for numbers,
```
"value": true
```
for booleans.

File: src/test/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBqInsertGeneratorTest.java
Comment (davidangb) [inline src/test/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBqInsertGeneratorTest.java:42]: see the comment on `PointCorrectionOperation` - these won't always be strings.

**Review by davidangb — COMMENTED** [review]
some more comments inline, most importantly that values in the inbound json files aren't always strings.

File: src/main/java/bio/terra/cloudfiletodatastore/GsonConverter.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/GsonConverter.java:10]: WFM

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBigQueryWriter.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBigQueryWriter.java:63]: It is for inserting larger amounts of data.  Here it doesn't matter.  I'm getting ahead of myself, removing.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerFileUploadedMessageProcessor.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerFileUploadedMessageProcessor.java:48]: Fair.  The content-type is only useful as long as it's in sync with the actual content and it's probably not a huge deal to fail processing with a JSON parsing error rather than skipping out upon a mismatched content-type.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/DeltaLayerBackgroundFunction.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/DeltaLayerBackgroundFunction.java:16]: Gson seems to be very forgiving.  Extra data in the JSON payload is ignored, missing data is deserialized as null.  Thus, there are fewer parse errors, but greater potential for downstream NPEs.  I've added a GcsEvent deserialization test confirming the expected behavior.

File: src/main/java/bio/terra/cloudfiletodatastore/FileUploadedMessage.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/FileUploadedMessage.java:72]: I'm going to push back and say I find this approach is easier for new devs to this code to understand.  Generating equals and hashcode with Intellij is a snap too should we need add more fields.

File: src/main/java/bio/terra/cloudfiletodatastore/GsonConverter.java
Comment (ichengchang) [reply src/main/java/bio/terra/cloudfiletodatastore/GsonConverter.java]: The `convertFromClass` method is now merged  into `GsonWrapper` in this [PR](https://github.com/DataBiosphere/terra-gcs-bq-streaming-functions/pull/11).

Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/GsonConverter.java]: It should be in GsonWrapper in the latest version of this branch too.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/DeltaLayerBackgroundFunction.java
Comment (ichengchang) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/DeltaLayerBackgroundFunction.java:19]: The `DeltaLayerBackgroundFunction` can just extend a `CloudStorageEventHarness<GCSEvent>` and put all the business logic in the `doAccept()` method. [QA-1544](https://github.com/DataBiosphere/terra-gcs-bq-streaming-functions/pull/11) has the `CloudStorageEventHarness`

`public abstract class CloudStorageEventHarness<T> implements BackgroundFunction<T>`

Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/DeltaLayerBackgroundFunction.java:19]: I don't see the advantage of adding the CloudStorageEventHarness as a superclass.   Am I missing something?

File: src/main/java/bio/terra/cloudfiletodatastore/FileUploadedMessage.java
Comment (davidangb) [reply src/main/java/bio/terra/cloudfiletodatastore/FileUploadedMessage.java:72]: that's fine - I don't feel strongly about this. If Immutables (or Lombok) become common in DSP we can always adopt it later, if it seems worthwhile.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBigQueryWriter.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBigQueryWriter.java]: Yes, this is definitely a weakness of this approach.  Big Query does have parameterized queries now but they can't be used in constructing the table name.  https://stackoverflow.com/questions/51475252/can-i-use-a-query-parameter-in-a-table-name/51479617  The values will always come from the DL JSON file, but I'll add a regex on project and data set name.

File: src/main/java/bio/terra/cloudfiletodatastore/FileUploadedMessage.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/FileUploadedMessage.java:86]: Definitely, I'm happy to adopt a standard, but my observation (hopefully not too harsh) after about 5 months at Broad is the standard is 'let's mix it up'.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBqInsertGenerator.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBqInsertGenerator.java:72]: TOL: this method _might possibly_ have perf issues under very heavy load:
* Boolean TRUEs, strings, dates and timestamps will call `strval.toLowerCase` twice, when possibly they could have called it once
* dates and timestamps will call `OffsetDateTime.parse` or `LocalDate.parse` twice, respectively, when possibly they could have called it once

Fixing this is probably premature optimization, but it's something to keep an eye on

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerFileUploadedMessageProcessor.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerFileUploadedMessageProcessor.java:18]: `EXPECTED_CONTENT_TYPE` is unused

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerFileUploadedMessageProcessor.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerFileUploadedMessageProcessor.java:22]: `bqForTest` is unused

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerFileUploadedMessageProcessor.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerFileUploadedMessageProcessor.java:59]: suggest adding a big try/catch around this and logging errors, or even adding more granular try/catches to the individual steps

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/GcsFileFetcher.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/GcsFileFetcher.java:29]: TODO either for this PR or for a future PR: add retries here to handle transient errors

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/DeltaLayerCloudFunction.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/DeltaLayerCloudFunction.java:17]: you'll sense the theme here: can this just be deleted in favor of `DeltaLayerBackgroundFunction`?

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/DeltaLayerRawFunction.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/DeltaLayerRawFunction.java:24]: delete in favor of `DeltaLayerBackgroundFunction`

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/MessageConverter.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/MessageConverter.java:6]: delete, if everything that calls this is also deleted

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/model/json/PointCorrectionRequest.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/model/json/PointCorrectionRequest.java:19]: this could be a `UUID` instead of `String`

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/model/json/PointCorrectionRequest.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfiletodatastore/deltalayer/model/json/PointCorrectionRequest.java:27]: same comment about `Instant` vs `OffsetDateTime`

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBigQueryWriter.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBigQueryWriter.java]: I've added an input check that hopefully does not create more problems than it solves.

File: src/main/java/bio/terra/cloudfunctions/common/GsonWrapper.java
Comment (davidangb) [inline src/main/java/bio/terra/cloudfunctions/common/GsonWrapper.java:34]: see https://github.com/DataBiosphere/terra-gcs-bq-streaming-functions/pull/11#discussion_r685492466 - this is not thread-safe

**Review by davidangb — COMMENTED** [review]
comments inline!

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBigQueryWriter.java
Comment (davidangb) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBigQueryWriter.java]: fwiw, here's what Rawls does: https://github.com/broadinstitute/rawls/blob/d3469764e50820e4147386dcac86083a73afa6b5/core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala#L28-L63

File: build.gradle
Comment (sagehen03) [reply build.gradle:26]: I agree.  One of my bullets in the PR description is we should choose the GCF approach we like best.  My preference is `DeltaLayerRawFunction`, yours is `DeltaLayerCloudFunction` and Ivan's is `DeltaLayerCloudFunction` :)

File: src/main/java/bio/terra/cloudevents/GCSEvent.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudevents/GCSEvent.java:17]: Yeah, no reason at all I think Ivan just pulled this code from the sample and added some fields in case we need them.  If we go with `DeltaLayerRawFunction` we don't need this class at all which is my main reason for supporting that approach.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBQSQLWriter.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBQSQLWriter.java:1]: see my above comment.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBQJSONWriter.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBQJSONWriter.java:36]: I think try with resources is out because we need the writer reference after close is called in order to do the error check (try with resources will cause the writer reference to disappear from scope after it's closed).  We could move the close to a finally block but it's pretty ugly..
```    
try {
        writer.write(ByteBuffer.wrap(jsonBytes));
      } catch (IOException e) {
        logger.log(Level.SEVERE, "Could not write to BQ", e);
        return;
      } finally {
        try {
          writer.close();
        } catch (IOException e) {
          logger.log(Level.SEVERE, "Could not close BQ writer", e);
        }
      }
```

File: src/main/java/bio/terra/cloudfiletodatastore/ResourceFetcher.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/ResourceFetcher.java:11]: I can just get rid of the default that's actually a vestige of having two methods.

File: src/main/java/bio/terra/cloudevents/GCSEvent.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudevents/GCSEvent.java:17]: Currently GCSEvent is still used by code that's in master so we can't delete just yet.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/MessageConverter.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/functions/MessageConverter.java:6]: I like having this logic separate so it can be tested apart from the GCF class.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBQSQLWriter.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBQSQLWriter.java:1]: I've deleted two of the three options from the PR.

File: src/main/java/bio/terra/cloudfunctions/common/GsonWrapper.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfunctions/common/GsonWrapper.java:34]: should be fixed now

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/model/json/PointCorrectionRequest.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/model/json/PointCorrectionRequest.java:28]: I think a zero offset OffsetDateTime is the same as an Instant.  I'm not sure sure there's harm here in using OffsetDateTime and there seems to be info suggesting it should be used when storing datetimes in BQ.  https://stackoverflow.com/questions/30234594/whats-the-difference-between-zoneddatetime-and-offsetdatetime

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerFileUploadedMessageProcessor.java
Comment (davidangb) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerFileUploadedMessageProcessor.java:18]: `EXPECTED_CONTENT_TYPE` and `bqForTest` below are both still unused

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerFileUploadedMessageProcessor.java
Comment (davidangb) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerFileUploadedMessageProcessor.java:59]: I'd still like to see some try/catch and logging of errors

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/GcsFileFetcher.java
Comment (davidangb) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/GcsFileFetcher.java:29]: would you mind adding a `// TODO:` comment here about retries?

**Review by davidangb — COMMENTED** [review]
a few comments inline, some holdovers from the previous review that weren't addressed.

Thanks for deleting 2 of the 3 entrypoint classes! We still have the choice of `DeltaLayerBQSQLWriter` vs `DeltaLayerBQJSONWriter`, and hopefully delete the one we're not using. I definitely prefer `DeltaLayerBQJSONWriter`, not least because we don't have to do our own SQL sanitization.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBQSQLWriter.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBQSQLWriter.java:57]: I believe underscores are included in "word" characters as far as java is concerned.  https://docs.oracle.com/javase/tutorial/essential/regex/pre_char_classes.html I'll add a unit test to be sure.

Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBQSQLWriter.java:57]: > couldn't you just include the errorCheck inside the try-with-resources?
I don't think so, `getJob()` returns null unless `close()` has been called first.  If you're still in the try with resources block `close()` wouldn't have been called yet, right?

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBigQueryWriter.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBigQueryWriter.java]: Wouldn't the Rawls validation allow `--` which denotes the beginning of a comment?  It's probably not a risk, but to be extra safe seems like that should excluded too.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBQSQLWriter.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBQSQLWriter.java:57]: Still pretty ugly, but maybe better than then try.. catch nested inside the finally block.

File: src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBqInsertGenerator.java
Comment (sagehen03) [reply src/main/java/bio/terra/cloudfiletodatastore/deltalayer/DeltaLayerBqInsertGenerator.java:71]: It seems like there's also a good chance this code will go away if we start writing the target data type in the JSON file.

**Review by davidangb — APPROVED** [review]
:+1:

* thanks for the clarification on supporting underscores in dataset names/regex - that's my bad, my comment was incorrect.
* the issue of try-with-resources and `.close()` is irrelevant now that `DeltaLayerBQJSONWriter` is gone
* unfortunately, this PR will have to merge into `main` instead of `master`, once #15 merges 

For future work - not blocking this PR - I suggest we evaluate creating the EAV table at dataset-creation time instead of creating it as needed at insertion time. Since we can't use the `CREATE_IF_NEEDED` disposition, I think the next-best option is to handle the table creation upfront (which means in Rawls, probably). This would also allow removing all the SQL-injection attack planes.

**Comment by sagehen03** [comment]
Lies! I didn't close this.  GH closed it when I tried to change the base branch to main.

**Pull request closed without merging**
